lang/LangPrimSource/SC_PortAudioPrim.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+    SuperCollider real time audio synthesis system
+    Copyright (c) 2002 James McCartney. All rights reserved.
+    http://www.audiosynth.com
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
+*/
+
+#include "portaudio.h"
+#include "SC_PaUtils.hpp"
+
+#include "SCBase.h"
+#include "VMGlobals.h"
+#include "PyrKernel.h"
+#include "PyrPrimitive.h"
+#include "GC.h"
+
+enum { OUT = 0, IN, BOTH };
+
+int listDevices(struct VMGlobals* g, int type) {
+    int numDevices, num = 0;
+    PyrSlot* a = g->sp - 2;
+
+    PaError err = Pa_Initialize();
+    if (err != paNoError)
+        return 0;
+
+    numDevices = Pa_GetDeviceCount();
+    if (numDevices < 0)
+        return 0;
+
+    if (type < BOTH) {
+        if (type < IN) {
+            for (PaDeviceIndex i = 0; i < numDevices; i++) {
+                auto* pdi = Pa_GetDeviceInfo(i);
+                if (pdi->maxOutputChannels > 0)
+                    num++;
+            }
+        } else {
+            for (PaDeviceIndex i = 0; i < numDevices; i++) {
+                auto* pdi = Pa_GetDeviceInfo(i);
+                if (pdi->maxInputChannels > 0)
+                    num++;
+            }
+        }
+    } else
+        num = numDevices;
+
+    PyrObject* devArray = newPyrArray(g->gc, num * sizeof(PyrObject), 0, true);
+    SetObject(a, devArray); // this is okay here as we don't use the receiver below
+
+    int i, j = 0;
+    for (i = 0; i < numDevices; i++) {
+        std::string name;
+        auto* pdi = Pa_GetDeviceInfo(i);
+        if (type < BOTH) {
+            if (type < IN) {
+                if (pdi->maxOutputChannels > 0)
+                    name = GetPaDeviceName(i);
+            } else {
+                if (pdi->maxInputChannels > 0)
+                    name = GetPaDeviceName(i);
+            }
+        } else
+            name = GetPaDeviceName(i);
+
+        if (!name.empty()) {
+            PyrString* string = newPyrString(g->gc, name.c_str(), 0, true);
+            SetObject(devArray->slots + j, string);
+            devArray->size++;
+            g->gc->GCWriteNew(devArray, (PyrObject*)string); // we know array is white so we can use GCWriteNew
+            j++;
+        }
+    }
```

Comment:
this function body could be written much more idiomatically. essentially, it's 3 steps:
1. collect device infos (which already contain the name of the device, btw; you don't need `GetPaDeviceName` here)
2. filter by the IN/OUT condition
3. transform filtered device infos into a PyrArray

the first step is just a `for` loop filling a `std::vector`; the second step you can write with the standard library functions [std::remove_if](https://en.cppreference.com/w/cpp/algorithm/remove) and [std::vector::erase](https://en.cppreference.com/w/cpp/container/vector/erase). you can use the size of the vector after filtering as a replacement for the variable `num` in this function body.

the third step is then also a simple `for` loop.

since device names are null-terminated `const char*`, you don't have to use an intermediary `std::string` to hold that data.

Reply:
OK, I'll try to implement the the logic you suggest.
However, we do need `GetPaDeviceName` - on Windows that returns `backend name : device name` and is consistent with what the server uses.

Reply:
ah okay, i forgot it does more than just read the name field on Windows. probably worth adding an overloading taking a PaDeviceInfo so you don't need to call PaGetDeviceInfo twice and keep around a list of indices

Reply:
an overload of GetPaDeviceName, i mean

Reply:
Do you mean for GetPaDeviceName to be able to take device info on top of device index?

Reply:
(answered on slack)